### PR TITLE
S3 bridge, stream improvements

### PR DESCRIPTION
**Pull request by pierotofy**
I'm starting an effort to improve stream handling for GET requests, since I've already implemented improvements in this area for the S3 bridge (a new controller that lets us access S3 resources locally via http).

### CODE REVIEW

File: Registry.Web/Models/DTO/FileStreamDescriptor.cs [hunk 1]
```diff
@@ -113,11 +113,6 @@ private async Task WriteObjectContentStream(string orgSlug, Guid internalRef, st
             if (!bucketExists)
                 throw new NotFoundException($"Cannot find bucket '{bucketName}'");
 
-            var objInfo = await _objectSystem.GetObjectInfoAsync(bucketName, path);
```
Comment (HeDo88TH) [inline Registry.Web/Models/DTO/FileStreamDescriptor.cs:116]: This check is required to ensure that the file actually exists in the bucket. It does not download the file, just assesses its presence.

**Comment by pierotofy** [comment]
Closes #200

**Comment by pierotofy** [comment]
Closes #203

**Comment by pierotofy** [comment]
Closes #204

**Comment by pierotofy** [comment]
Note on the S3 bridge: if there's a reverse-proxy in front of Registry, it's likely that those requests are coming from localhost, so this would be a security problem (mitigated by the fact that the bucket UUIDs are hidden, but still).

The reverse proxy should be configured to block all paths that begin with `_` (e.g. `http://hub.dronedb.app/_bridge`) since they are for internal use only.

**Comment by pierotofy** [comment]
#204 was actually already fixed; I just rewrote the logic a bit I guess.

**Comment by pierotofy** [comment]
Ready for review @HeDo88TH

File: Registry.Adapters/ObjectSystem/CachedS3ObjectSystem.cs [hunk 2]
```diff
@@ -309,8 +309,7 @@ private void UpdateCurrentCacheSize()
         public void Cleanup()
         {
             CleanupFolder(CachePath);
-
-            Directory.EnumerateDirectories(CachePath).ToList().ForEach(CommonUtils.RemoveEmptyFolders);
+            CommonUtils.RemoveEmptyFolders(CachePath);
```
Comment (pierotofy) [inline Registry.Adapters/ObjectSystem/CachedS3ObjectSystem.cs:312]: fyi: I refactored, it's not a regression.

**Comment by pierotofy** [comment]
When S3 is enabled, we can now access from localhost:

`GET http://localhost:5000/_bridge/{bucket}/{path}`
`HEAD http://localhost:5000/_bridge/{bucket}/{path}`

Http range requests for partial downloads are supported also.

**Comment by pierotofy** [comment]
EPT thumbnails are next.

**Comment by pierotofy** [comment]
Note on the physical file system implementation: we no longer close the file streams on GetObjectAsync. It's responsibility of the caller to do so. I think the unit tests don't care about this, but calls elsewhere might.

File: Registry.Adapters/ObjectSystem/CachedS3ObjectSystem.cs (hunk 2 above)
Comment (HeDo88TH) [reply Registry.Adapters/ObjectSystem/CachedS3ObjectSystem.cs:312]: 👍

**Comment by HeDo88TH** [comment]
I tested it on windows and It gave me an error in the `Build` endpoint.

There is a much simpler way to implement it without using `TaskCompletionSource`. 
Check https://github.com/DroneDB/Registry/pull/199/commits/1e7b4f86890fb70dd7be9b9f31e3e9d22a7b0de9. 
We should use the same tecnique when we need to stream any content without intermediate buffers (aka `MemoryStream`).
It was already used in the download endpoint. I created a generic class `StreamableFileDescriptor` that is more versatile.

Here it is a simple stress-test I run using `G_Sw_Anny.laz`

getbuildfile-profiled

The memory usage goes up but then goes down as it correctly releases the streams after each burst.

**Comment by pierotofy** [comment]
`There is a much simpler way to implement it without using TaskCompletionSource`.

To my understanding, `CopyTo` does not let you "link" the streams, but copies all data immediately. Have you plotted memory usage using my proposed approach?

**Comment by HeDo88TH** [comment]
> `There is a much simpler way to implement it without using TaskCompletionSource`.
> To my understanding, `CopyTo` does not let you "link" the streams, but copies all data immediately. 

Yes but in my approach I'm writing directly on the Response.Body stream, nothing is stored in memory.

> Have you plotted memory usage using my proposed approach and checked how it compares?

It crashed throwing a NullReferenceException. I tried serveral times with different storage systems but I could not make it work.

> We're just downloading files; there should be **no** memory spikes, ever.

Every spike is ~1000 requests and it contains all the DB access and the other things that ASP.NET does per request. The usage behaviour is fine!

**Comment by pierotofy** [comment]
Just write in this thread what steps you did to crash it. I will fix the issue so we can run a comparison test.

**Comment by pierotofy** [comment]
I ran some tests; performance memory-wise is the same between the two methods. You just never know until you test. Not sure how you got the null reference, but doesn't matter. The `StreamableFileDescriptor` is nice.

Btw, seems like using pure streams for high performance I/O are not recommended, at some point in the future we might want to look into IO.Pipelines. https://devblogs.microsoft.com/dotnet/system-io-pipelines-high-performance-io-in-net/

**Comment by HeDo88TH** [comment]
> I ran some tests; performance memory-wise is the same between the two methods. You just never know until you test. Not sure how you got the null reference, but doesn't matter. The `StreamableFileDescriptor` is nice.
> 
> Btw, seems like using pure streams for high performance I/O is not recommended, at some point in the future we might want to look into IO.Pipelines. https://devblogs.microsoft.com/dotnet/system-io-pipelines-high-performance-io-in-net/

I knew that article but I never had the chance to use System.IO.Pipelines. I think the first step is to close https://github.com/DroneDB/Registry/issues/197 by removing all the intermediate memory buffers.

**Pull request merged**